src/main/java/com/softserve/academy/spaced/repetition/controller/CardsFileController.java

```diff
@@ -0,0 +1,62 @@
+package com.softserve.academy.spaced.repetition.controller;
+
+import com.softserve.academy.spaced.repetition.exceptions.EmptyFileException;
+import com.softserve.academy.spaced.repetition.exceptions.NotAuthorisedUserException;
+import com.softserve.academy.spaced.repetition.exceptions.NotOwnerOperationException;
+import com.softserve.academy.spaced.repetition.exceptions.WrongFormatException;
+import com.softserve.academy.spaced.repetition.service.CardService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.util.FileCopyUtils;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+import javax.servlet.http.HttpServletResponse;
+import java.io.*;
+
+@Controller
+public class CardsFileController {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CardsFileController.class);
+
+    @Autowired
+    private CardService cardService;
+
+    @PostMapping("api/private/upload/deck/{deckId}/cards")
+    public ResponseEntity uploadFile(@RequestParam("file") MultipartFile cardsFile, @PathVariable Long deckId) throws WrongFormatException, NotOwnerOperationException, NotAuthorisedUserException, EmptyFileException, IOException {
+        cardService.uploadCards(cardsFile, deckId);
+        return new ResponseEntity(HttpStatus.OK);
+    }
+
+    @GetMapping("download/deck/{deckId}/cards")
+    public StreamingResponseBody downloadFile(HttpServletResponse response, @PathVariable Long deckId) {
+        response.setContentType("application/octet-stream");
+        response.setHeader("Content-Disposition", "attachment; filename=Cards.yml");
+        return outputStream -> {
```

Comment:
You have added a lib org.yaml:snakeyaml, to work with yaml.
On the other hand you are converting "card" to yaml by yourself.

The best practice would be to use a functionality that is already implemented if it is available on your classpath. There is a method [Yaml#dump](http://javadox.com/org.yaml/snakeyaml/1.13/org/yaml/snakeyaml/Yaml.html#dump(java.lang.Object, java.io.Writer)). With a StringWriter you can convert it to String and then to bytes.

Would be also nice to move the conversion (or its part) from controller to service layer.

Would be also nice to add a unit test. :)

Reply:
@antonbakalets that is, collect all in one big string and dump?

Reply:
No, collect all in one object (a new dto?) then dump to string or to outputstream if possible.